ArduPlane/mode_autoland.cpp

```diff
@@ -129,30 +139,97 @@ void ModeAutoLand::update()
+    switch (stage) {
+    case AutoLandStage::LOITER:
+        // check if we have arrived and completed loiter at base leg waypoint
+        if (plane.verify_loiter_to_alt(cmd[0])) {
+            stage = AutoLandStage::BASE_LEG;
+            plane.start_command(cmd[1]);
+            // setup a predictable turn angle for the base leg
+            plane.auto_state.next_turn_angle = 90;
+            plane.auto_state.crosstrack = false;
+        }
+        break;
+
+    case AutoLandStage::BASE_LEG:
+        // check if we have reached the final approach waypoint
+        if (plane.verify_nav_wp(cmd[1])) {
+            stage = AutoLandStage::LANDING;
+            plane.start_command(cmd[2]);
+        }
+        break;
+
+    case AutoLandStage::LANDING:
         plane.set_flight_stage(AP_FixedWing::FlightStage::LAND);
-        plane.verify_command(cmd[stage]);
+        plane.verify_command(cmd[2]);
+        // make sure we line up
+        plane.auto_state.crosstrack = true;
+        break;
+    }
+}
+
+/*
+  Takeoff direction is initialized after arm when sufficient altitude
+  and ground speed is obtained, then captured takeoff direction +
+  offset used as landing direction in AUTOLAND
+*/
+void ModeAutoLand::check_takeoff_direction()
+{
+#if HAL_QUADPLANE_ENABLED
+    // we don't allow fixed wing autoland for quadplanes
+    if (quadplane.available()) {
         return;
     }
+#endif
 
-    // see if we have completed the leg
-    if (plane.verify_nav_wp(cmd[stage])) {
-        stage++;
-        plane.prev_WP_loc = plane.next_WP_loc;
-        plane.auto_state.next_turn_angle = 90;
-        plane.start_command(cmd[stage]);
-        plane.auto_state.crosstrack = true;
-        plane.auto_state.next_wp_crosstrack = true;
+    if (plane.takeoff_state.initial_direction.initialized) {
+        return;
+    }
+    //set autoland direction to GPS course over ground
+    if (plane.control_mode->allows_autoland_direction_capture() &&
+        plane.is_flying() &&
+        hal.util->get_soft_armed() &&
+        plane.gps.ground_speed() > GPS_GND_CRS_MIN_SPD) {
+        set_autoland_direction();
     }
 }
-#endif
+
+// Sets autoland direction using ground course + offest parameter
+void ModeAutoLand::set_autoland_direction()
+{
+    plane.takeoff_state.initial_direction.heading = wrap_360(plane.gps.ground_course() + landing_dir_off);
+    plane.takeoff_state.initial_direction.initialized = true;
+    gcs().send_text(MAV_SEVERITY_INFO, "Autoland direction= %u",int(plane.takeoff_state.initial_direction.heading));
+}
+
+/*
+  return true when the LOITER_TO_ALT is lined up ready for the landing, used in commands_logic verify loiter to alt
+ */
+bool ModeAutoLand::landing_lined_up(void)
+{
+    // use the line between the center of the LOITER_TO_ALT on the base leg and the
+    // start of the landing leg (land_start_WP)
+    return plane.mode_loiter.isHeadingLinedUp_cd(cmd[0].content.location.get_bearing(cmd[1].content.location)*100);
```

Comment:
Why not use `isHeadingLinedUp(cmd[0].content.location, cmd[1].content.location)`? It has lots of extra handling that should make it more reliable.

Reply:
It should also better deal with the case that the loiter radius is larger than the base leg.

Reply:
I had that and Tridge did not like the way it behaved